### PR TITLE
update the query parameters and refactoring of code.

### PR DESCRIPTION
**Pull request by Pradeepkumarbk**
This PR updates the query parameters and also fixes the failure issues.

Signed-off-by: Pradeep <pradeepkumar.bk96@gmail.com>

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -1,374 +1,357 @@
...
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var URL string
+var ClientSet *kubernetes.Clientset
+
+type PVMetrics struct {
+	ReadIops        float64
+	writeIops       float64
```
Comment (Akash4927) [inline main.go]: It should be WriteIops

File: main.go [hunk 2]
```diff
@@ -1,374 +1,357 @@
...
+type report struct {
+	PersistentVolume topology
+	Plugins          []pluginSpec
+}
+
+type response struct {
+	ShortcutReport *report `json:"shortcutReport,omitempty"`
+}
+
+type KubernetesAPI struct {
```
Comment (Akash4927) [inline main.go]: Is this required?

File: main.go [hunk 3]
```diff
@@ -1,374 +1,357 @@
...
+
+func (p *Plugin) makeReport() (*report, error) {
+	go p.updateIops()
+	go p.updateLatency()
+	go p.updateThroughput()
+
+	metrics := make(map[string][]float64)
+	resource := make(map[string]node)
+
+	for Iops, iopsMetrics := range p.Iops {
```
Comment (Akash4927) [inline main.go]: Iops should be pvUID, because the map key will be the UID of PV

File: main.go [hunk 4]
```diff
@@ -1,374 +1,357 @@
...
+		metrics[p.getPVTopology(Iops)] = append(metrics[p.getPVTopology(Iops)], iopsMetrics.ReadIops, iopsMetrics.writeIops)
+	}
+
+	for Latency, latencyMetrics := range p.Latency {
```
Comment (Akash4927) [inline main.go]: Latency should be pvUID, because the map key will be the UID of PV

File: main.go [hunk 5]
```diff
@@ -1,374 +1,357 @@
...
+		metrics[p.getPVTopology(Latency)] = append(metrics[p.getPVTopology(Latency)], latencyMetrics.ReadLatency, latencyMetrics.WriteLatency)
+	}
+
+	for Throughput, throughputMetrics := range p.Throughput {
```
Comment (Akash4927) [inline main.go]: Throughput should be pvUID, because the map key will be the UID of PV

File: main.go [hunk 6]
```diff
@@ -1,374 +1,357 @@
...
+func (p *Plugin) makeReport() (*report, error) {
+	go p.updateIops()
+	go p.updateLatency()
+	go p.updateThroughput()
+
+	metrics := make(map[string][]float64)
+	resource := make(map[string]node)
+
+	for Iops, iopsMetrics := range p.Iops {
+		metrics[p.getPVTopology(Iops)] = append(metrics[p.getPVTopology(Iops)], iopsMetrics.ReadIops, iopsMetrics.writeIops)
```
Comment (Akash4927) [inline main.go]: writeIops should be WriteIops

File: main.go [hunk 7]
```diff
@@ -1,374 +1,357 @@
...
+		"writeIops": {
+			ID:       "writeIops",
+			Label:    "Iops(W)",
+			Format:   "",
+			Priority: 0.2,
+		},
+		"readLatency": {
+			ID:       "readLatency",
+			Label:    "Latency(R)",
+			Format:   "",
```
Comment (Akash4927) [inline main.go]: Why format is empty for Latency and throughput? 
I remember, you raised a PR for adding the format in Scope.

File: throughput.go [hunk 8]
```diff
@@ -1,87 +1,86 @@
...
-	queries := []string{"OpenEBS_read_block_count_per_second", "OpenEBS_write_block_count_per_second"}
+func getThroughputMetrics() map[string]PVMetrics {
+	queries := []string{throughputReadQuery, throughputWriteQuery}
 	for _, query := range queries {
-		go getTputValues(url+query, query)
+		go getThroughputValues(query)
 	}
-	var readThroughput, writeThroughput map[string]float64
+
+	readthroughput := make(map[string]float64)
```
Comment (Akash4927) [inline throughput.go]: readthroughput and writethroughput should be in camelCase

File: iops.go [hunk 9]
```diff
@@ -1,87 +1,86 @@
...
-	queries := []string{"OpenEBS_read_iops", "OpenEBS_write_iops"}
+func getIopsMetrics() map[string]PVMetrics {
+	queries := []string{iopsReadQuery, iopsWriteQuery}
 	for _, query := range queries {
-		go getValues(url+query, query)
+		go getValues(query)
 	}
-	var read, write map[string]float64
+
+	readiops := make(map[string]float64)
```
Comment (Akash4927) [inline iops.go]: It should be in camelCase

File: latency.go [hunk 10]
```diff
@@ -1,87 +1,86 @@
...
-	queries := []string{"OpenEBS_read_latency", "OpenEBS_write_latency"}
+func getLatencyMetrics() map[string]PVMetrics {
+	queries := []string{latencyReadQuery, latencyWriteQuery}
 	for _, query := range queries {
-		go getLatValues(url+query, query)
+		go getLatencyValues(query)
 	}
-	var readLatency, writeLatency map[string]float64
+
+	readlatency := make(map[string]float64)
```
Comment (Akash4927) [inline latency.go]: camelCase here

File: iops.go [hunk 11]
```diff
@@ -1,87 +1,86 @@
...
 	"net/http"
 	"strconv"
-	"time"
 
+	log "github.com/sirupsen/logrus"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-var pv = make(map[string]float64)
+var iopsReadChan = make(chan map[string]float64)
```
Comment (Akash4927) [inline iops.go]: Can you group the variables like 
```
var (
       iopsReadChan = make(chan map[string]float64)
       iopsWriteChan = make(chan map[string]float64)
)
```

File: iops.go [hunk 12]
```diff
@@ -1,87 +1,86 @@
...
+var iopsWriteChan = make(chan map[string]float64)
 
-type pvdata struct {
-	read  float64
-	write float64
-}
+var iopsReadQuery = "increase(openebs_reads[5m])/300"
```
Comment (Akash4927) [inline iops.go]: Similar here and in latency and throughput.go files too.

**Comment by Akash4927** [comment]
Please update the test files too.

File: iops_test.go [hunk 13]
```diff
@@ -48,47 +48,22 @@ func TestGetValues(t *testing.T) {
 	for name, tt := range cases {
 		t.Run(name, func(t *testing.T) {
 			server := httptest.NewServer(&tt.fakeHandler)
-			go getValues(server.URL, tt.queryType)
+			URL = server.URL
+			go getValues(tt.queryType)
```
Comment (satyamz) [inline iops_test.go]: Can you explain me here what this test does? (As you've not mentioned in the commit message) Also, Do we need `go` routine in the test? :thinking:

File: latency.go [hunk 14]
```diff
@@ -1,87 +1,90 @@
...
-	writeLatency float64
-}
+var (
+	latencyReadQuery  = "increase(openebs_read_time[5m])/1000000"
+	latencyWriteQuery = "increase(openebs_write_time[5m])/1000000"
+)
 
-func getLatValues(urlpassed string, query string) {
-	res, err := http.Get(urlpassed)
+func getLatencyValues(query string) {
```
Comment (satyamz) [inline latency.go]: Is there a way to combine the code? I see these methods are similar to each. You can tie these methods to the struct which has these parameters in it and method to work on those parameter.

File: iops_test.go (hunk 13 above)
Comment (Pradeepkumarbk) [reply iops_test.go]: yes, we need go routine because the function that we are testing returns the output in channel.

Comment (satyamz) [reply iops_test.go]: I think, first you should write a test which tests `getValues` method first.

**Pull request closed without merging**